src/util/mod.rs

```diff
@@ -53,3 +54,95 @@ pub fn make_spirv<'a>(data: &'a [u8]) -> super::ShaderModuleSource<'a> {
+                if mapped {
+                    map_context.initial_range = 0..size;
+                }
+                crate::Buffer {
+                    context: Arc::clone(&self.context),
+                    id: crate::Context::device_create_buffer(&*self.context, &self.id, &wgt_descriptor),
+                    map_context: parking_lot::Mutex::new(map_context),
+                    usage: descriptor.usage,
+                }
+            },
+            BufferContents::CopyData(data) => {
+                let unpadded_size = data.len() as crate::BufferAddress;
+                let padding = crate::COPY_BUFFER_ALIGNMENT - unpadded_size % crate::COPY_BUFFER_ALIGNMENT;
+                let padded_size = padding + unpadded_size;
+
+                let wgt_descriptor = wgt::BufferDescriptor {
+                    label: descriptor.label.as_ref().map(|x| Cow::Borrowed(&**x)),
+                    size: padded_size,
+                    usage: descriptor.usage,
+                    mapped_at_creation: true,
+                };
+
+                let mut map_context = crate::MapContext::new(padded_size);
+
+                map_context.initial_range = 0..padded_size;
+                let buffer = crate::Buffer {
+                    context: Arc::clone(&self.context),
+                    id: crate::Context::device_create_buffer(&*self.context, &self.id, &wgt_descriptor),
+                    map_context: parking_lot::Mutex::new(map_context),
+                    usage: descriptor.usage,
+                };
+
+                let range =
+                    crate::Context::buffer_get_mapped_range_mut(&*self.context, &buffer.id, 0..padded_size);
+                range[0..unpadded_size as usize].copy_from_slice(data);
+                for i in unpadded_size..padded_size {
+                    range[i as usize] = 0;
+                }
+
+                buffer.unmap();
+                buffer
+            }
+        }
+    }
+}
+
+/// Describes the contents of a [`Buffer`] at creation.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum BufferContents<'a> {
+    /// Allocate memory, but do not provide values for the contents.
+    Allocation {
+        /// The size of the allocation, in bytes.
+        size: crate::BufferAddress,
+        /// Allows a buffer to be mapped immediately after it is made. The buffer does not have to be [`BufferUsage::MAP_READ`]
+        /// or [`BufferUsage::MAP_WRITE`], all buffers are allowed to be mapped at creation.
+        mapped: bool,
+    },
+    /// Copy the values from the slice into the [`Buffer`] on creation.
+    CopyData(&'a [u8]),
+}
+
+/// Describes a [`Buffer`] when allocating.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct BufferInitDescriptor<'a, 'data> {
+    /// Debug label of a buffer. This will show up in graphics debuggers for easy identification.
+    pub label: Option<Cow<'a, str>>,
+    /// Contents of a buffer on creation.
+    pub contents: BufferContents<'data>,
```

Comment:
Sorry for not being clear about this: we want `data: &'a [u8]` here instead of the contents enum. If the user is not providing the data, they should use the original `Device::create_buffer`.